### DISCUSSION TITLE
Spawning multiple jobs in a loop

### DISCUSSION THREAD

**Discussion by ottramst** [go-co-op/gocron#265] [Q&A] [answered]
Hey guys,

Thanks for the awesome project!

Just some background - I'm a bit of a newbie when it comes to goroutines, channels and all that magic. So bare with me here.
I have a small question about "spawning" jobs in a loop. 

My end goal is to spawn multiple jobs based on an array (doesn't matter - strings, int's etc):

```
func main() {

	a := []int{1, 2, 3}

	s := gocron.NewScheduler(time.UTC)

	for _, i := range a {
		s.Every("1s").Do(func() { fmt.Println(i) })
	}

	s.StartBlocking()
}
```

This gives me output
```
3
3
3
...
```

So it seems that only the last of the loop is actually passed to the scheduler. Any advice here?

Thanks!

**Comment by JohnRoesler** [comment, marked as answer]
Hi @ottramst glad you are enjoying the project! And we were all there once (as newbies)! 😁 

This one is definitely sneaky and has to do with how the range values work in Go. Because you're passing in a function `func() { fmt.Println(i) }` the value of `i` isn't resolved immediately and when it is the value is 3, the final value in your loop. You can make it work the way you are desiring by copying the iterator value or passing it as a parameter to your function:

### copy iterator

```golang
package main

import (
	"fmt"
	"log"
	"time"

	"github.com/go-co-op/gocron"
)

func main() {

	a := []int{1, 2, 3}

	s := gocron.NewScheduler(time.UTC)

	for _, i := range a {
		j := i
		_, err := s.Every("1s").Do(func() { fmt.Println(j) })
		if err != nil {
			log.Fatalln(err)
		}
	}

	s.StartBlocking()
}
```

### as func param

```golang
package main

import (
	"fmt"
	"log"
	"time"

	"github.com/go-co-op/gocron"
)

func main() {

	a := []int{1, 2, 3}

	s := gocron.NewScheduler(time.UTC)

	for _, i := range a {
		_, err := s.Every("1s").Do(func(j int) { fmt.Println(j) }, i)
		if err != nil {
			log.Fatalln(err)
		}
	}

	s.StartBlocking()
}
```

**Comment by ottramst** [comment]
Thank you for this simple explanation!

**Reply by JohnRoesler** [reply]
You're very welcome 👍
